test/integration/test_cluster.c

```diff
@@ -97,22 +97,48 @@ TEST(cluster, restart, setUp, tearDown, 0, cluster_params)
 	struct rows rows;
 	long n_records =
 	    strtol(munit_parameters_get(params, "num_records"), NULL, 0);
-	char sql[128];
+	int rv;
 
 	HANDSHAKE;
 	OPEN;
 	PREPARE("CREATE TABLE test (n INT)", &stmt_id);
 	EXEC(stmt_id, &last_insert_id, &rows_affected);
 
+	PREPARE("INSERT INTO TEST(n) VALUES(?)", &stmt_id);
 	for (int i = 0; i < n_records; ++i) {
-		sprintf(sql, "INSERT INTO test(n) VALUES(%d)", i + 1);
-		PREPARE(sql, &stmt_id);
-		EXEC(stmt_id, &last_insert_id, &rows_affected);
+		struct value val = {.type = SQLITE_INTEGER, .integer = i};
+		rv = clientSendExec(f->client, stmt_id, &val, 1, NULL);
+		munit_assert_int(rv, ==, 0);
+		rv = clientRecvResult(f->client, &last_insert_id,
+				      &rows_affected, NULL);
+		munit_assert_int(rv, ==, 0);
 	}
 
 	struct test_server *server = &f->servers[0];
 	test_server_stop(server);
-	test_server_start(server, params);
+
+	test_server_prepare(server, params);
+	uint64_t last_entry_index;
+	uint64_t last_entry_term;
+	rv = dqlite_node_describe_last_entry(server->dqlite,
+					     &last_entry_index,
+					     &last_entry_term);
+	munit_assert_int(rv, ==, 0);
+	/* When we've inserted 0 records, there are two entries:
+	 * the initial configuration and the CREATE TABLE transaction.
+	 *
+	 * At 993 records, the leader has generated one legacy checkpoint
+	 * command entry.
+	 *
+	 * At 2200 records, the leader has generated a second checkpoint
+	 * command, plus two barriers. (The number of barriers is nondeterministic;
+	 * but we seem to get two of them pretty consistently in this test.) */
+	size_t fudge = n_records >= 2200 ? 6 :
+			n_records >= 993 ? 3 :
+			2;
```

Comment:
This seems a bit brittle especially the part about the barriers which are non-deterministic which could lead to CI failures down the line. What about changing the assertion so that if n_records < 993 then,  `last_entry_index <= n_records + 3 && last_entry_index >= n_records`? (or something similar, the idea is to use an interval and not a single value)

Reply:
I've modified this to use the calculated number just as a lower bound. Feels to me like writing something like `last_entry_index <= n_records + CONSTANT` is not much better than picking an exact value because we don't have an explanation for the specific `CONSTANT`; I would rather just have the assertion reflect our belief about the code i.e. "some number of barriers, we don't know how many".